miscellaneous/audit-service/service.bal

```diff
@@ -0,0 +1,148 @@
+import ballerina/http;
+import ballerinax/health.fhir.r4.international401;
+import ballerinax/health.fhir.r4;
+import ballerina/io;
+import ballerina/log;
+import ballerina/cache;
+import ballerina/uuid;
+import ballerinax/health.fhir.r4.terminology;
+import ballerina/task;
+
+// in Choreo context, this is expected to be a path in a file mount 
+configurable string auditLogPath = "/tmp/audit-logs/fhir-audit.log";
+// capacity of the cache used to store the failed audit events till they are retried
+configurable int cacheCapacity = 1000;
+// name of the fhir server. This is used as the source observer name in the FHIR audit event
+configurable string fhirServerName = "wso2fhirserver.com";
+// agent type of the audit event. This is used as the agent type in the FHIR audit event
+configurable string agentType = "humanuser";
+
+// This creates a new cache with the advanced configuration.
+final cache:Cache cache = new ({
+    capacity: cacheCapacity
+});
+
+// Retry failed audit events
+class RetryFailedAuditEvents {
+
+    *task:Job;
+
+    public function init() {
+        log:printDebug("Initialized the `retry failed audit events` task.");
+    }
+
+    // Executes this function when the scheduled trigger fires.
+    public function execute() {
+        int i = 0;
+        if cache.keys().length() > 0 {
+            log:printDebug("Retrying to write failed audit events to the log file.", numberOfFailedAuditEvents = cache.keys().length());
+        }
+        while i < cache.keys().length() {
+            // retry to write to the audit log file
+            international401:AuditEvent|error auditEvent = cache.get(cache.keys()[i]).ensureType();
+            if (auditEvent is international401:AuditEvent) {
+                io:Error? result = io:fileWriteLines(auditLogPath, [auditEvent.toJsonString()], option = io:APPEND);
+                if !(result is io:Error) {
+                    // if retrying is successful, remove from the cache
+                    check cache.invalidate(cache.keys()[i]);
+                    log:printDebug("Successfully wrote the audit event to the log file.", id = auditEvent.id);
+                } else {
+                    i += 1;
```

Comment:
If we want to retry, shouldn't we not increase i? Otherwise, the entry will still be in the cache but the loop will go to the next key?

Reply:
Also, wouldn't using a lambda instead of a while loop to traverse the cache be convenient?
cache.keys().forEach(function (string key) {});

Reply:
No, here the logic is a bit different. This logic runs as part of a periodic task. So, if there are entries in the cache, retrying for those entries will again happen in the next task cycle. And since we have to invalidate the cache key if it's written successfully, I'm not sure about the thread safety of lambda.